api/api.go

```diff
@@ -82,33 +90,72 @@ func (api *API) ListenAndServe(port int) error {
 	return http.ListenAndServe(fmt.Sprintf(":%d", port), api.staticRouter)
 }
 
-// WithDBSession injects a session context into the request context of the handler.
+// WithDBSession injects a session context into the request context of the
+// handler. In case of a MongoDB WriteConflict error, the call is retried up to
+// dbTxnRetryCount times or until the request context expires.
 func (api *API) WithDBSession(h httprouter.Handle) httprouter.Handle {
 	return func(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-		// Create a new db session
-		sess, err := api.staticDB.NewSession()
-		if err != nil {
-			api.WriteError(w, errors.AddContext(err, "failed to start a new mongo session"), http.StatusInternalServerError)
-			return
+		numRetriesLeft := dbTxnRetryCount
+		var body []byte
+		var err error
+		if req.Body != nil {
+			// Read the request's body and replace its Body io.ReadCloser with a
+			// new one based off the read data.
+			body, err = io.ReadAll(io.LimitReader(req.Body, LimitBodySizeLarge))
+			if err != nil {
+				api.WriteError(w, errors.AddContext(err, "failed to read body"), http.StatusBadRequest)
+				return
+			}
+			_ = req.Body.Close()
 		}
-		// Close session after the handler is done.
-		defer sess.EndSession(req.Context())
-
-		// Create session context.
-		sctx := mongo.NewSessionContext(req.Context(), sess)
 
-		// Get the special response writer.
-		mw, err := NewMongoWriter(w, sctx, api.staticLogger)
-		if err != nil {
-			api.WriteError(w, errors.AddContext(err, "failed to start a new transaction"), http.StatusInternalServerError)
+		for {
+			req.Body = io.NopCloser(bytes.NewReader(body))
+			// Create a new db session
+			sess, err := api.staticDB.NewSession()
+			if err != nil {
+				api.WriteError(w, errors.AddContext(err, "failed to start a new mongo session"), http.StatusInternalServerError)
+				return
+			}
+			// Close session after the handler is done.
+			defer sess.EndSession(req.Context())
```

Comment:
If we us defers like this in a loop, aren't we overwritting the defer call and not actually closing the previous calls? i.e. only the lass `defer` session is closed or the last session is closed multiple times because `sess` is overwritten?

Instead, should we refactor this code into a function that does fully close out the session each time before the next retry?

Reply:
sounds like this is fine.

fwiw my thought for containing it was simply something like
```go
for {
  err = func()
 // check err and decide to retry or not

}


func() {
 create session
 defer close session
}
```

fine to resolve.